tuned/plugins/plugin_intel_uncore.py

```diff
@@ -0,0 +1,174 @@
+from . import hotplug
+from .decorators import *
+import tuned.logs
+from tuned.utils.commands import commands
+
+import os
+import fnmatch
+
+log = tuned.logs.get()
+cmd = commands()
+
+SYSFS_DIR = "/sys/devices/system/cpu/intel_uncore_frequency/"
+
+IS_MIN = 0
+IS_MAX = 1
+
+class IntelUncorePlugin(hotplug.Plugin):
+	"""
+	`intel_uncore`::
+
+	`max_freq_khz, min_freq_khz`:::
+	Limit the maximum and minumum uncore frequency.
+	====
+	----
+	[intel_uncore]
+	max_freq_khz=4000000
+	----
+	Using this options *TuneD* will limit maximum frequency of all uncore units
+	in the system to 4 GHz
+	====
+	"""
+
+	def _init_devices(self):
+		self._devices_supported = True
+		self._assigned_devices = set()
+		self._free_devices = set()
+		self._is_tpmi = False
+
+		try:
+			devices = os.listdir(SYSFS_DIR)
+		except OSError:
+			return
+
+		# For new TPMI interface use only uncore devices
+		tpmi_devices = fnmatch.filter(devices, 'uncore*')
```

Comment:
Maybe you could use udev for listing of the devices? I.e. similar way how the cpu or disk plugin does. In such way the hotplug will also start working. It seems udev knows about uncores, but I wasn't able to deeper test it. I did very simple udev query for all available devices on my system and I can see:
```
...
Device('/sys/devices/uncore_arb')
Device('/sys/devices/uncore_cbox_0')
Device('/sys/devices/uncore_cbox_1')
Device('/sys/devices/uncore_imc')
...
```

Reply:
Hi, thanks for review!

> Maybe you could use udev for listing of the devices? I.e. similar way how the cpu or disk plugin do. In such 
case the hotplug will also start working.

Uncore frequency devices are just entries in sysfs directory, not "class" devices that generate udev events out of the box and have udev DEVPATH= . The intel_uncore_frequency.ko drivers support hotplug in that sense it will create sysfs entries when new physical processor is added or remove them when processor box is removed, however FWICT it will not generate any user-space events. I think this have to be improved on the driver side or at least validated.

> It seems udev knows about uncores, but I wasn't able to deeper test it. I did very simple udev query for all available devices on my system and I got:
> 
> ```
> ...
> Device('/sys/devices/uncore_arb')
> Device('/sys/devices/uncore_cbox_0')
> Device('/sys/devices/uncore_cbox_1')
> Device('/sys/devices/uncore_imc')
> ...
> ```

Those are performance monitoring devices, not strictly related to uncore_frequency.